docs/src/adr/015adr-trace.md

```diff
@@ -0,0 +1,403 @@
+  "vars":  <array of strings>,
+  "states": <array of states>,
+  "loop": <optional int>
+}
+```
+
+The field `#meta` is an arbitrary JSON object, whose purpose is to provide
+the reader with additional comments about the trace. For example, it may look
+like:
+
+```js
+  "#meta": {
+    "description": "Generated by Apalache",
+    "source": "MissionariesAndCannibalsTyped.tla"
+  }
+```
+
+The optional field `params` is an array of names that must be set in the
+initial state (if there are any parameters). The parameters play the same role
+as `CONSTANTS` in TLA+. For example, the field may look like:
+
+```js
+  "params": [ "Missionaries", "Cannibals" ]
+```
+
+The field `vars` is an array of names that must be set in every state.
+For example, the field may look like:
+
+```js
+  "vars": [ "bank_of_boat", "who_is_on_bank" ]
+```
+
+The field `states` is an array of state objects (see below). For example,
+the field may look like:
+
+```js
+  "states": [ <state0>, <state1>, <state2> ]
+```
+
+The optional field `loop` specifies the index of the state (in the array of
+states) that starts the loop. The loop ends in the last state. For example,
+the field may look like:
+
+```js
+  "loop": 1
+```
+
+**State object**. A state is a JSON object:
+
+```js
+  {
+    "#meta": <optional object",
+    "<var1>": <expr>,
+    ...
+    "<varN>": <expr>
+  }
+```
+
+As in the trace object, the field `#meta` may be an arbitrary object.
+Different tools may use this object to write their metadata into this object.
+
+The names `<var1>, ..., <varN>` are the names of the variables that are
+specified in the field `vars`. The syntax of `<expr>` is specified below.
+
+**Expressions.** As usual, expressions are inductively defined. An expression
+`<expr>` is one of the following:
+
+ 1. A JSON string literal, e.g., `"hello"`. TLA+ strings are written as strings
+    in this format.
+
+ 1. A JSON integer, e.g., 123. According to [RFC7159][], JSON integers
+    must be in the range: `[-(2**53)+1, (2**53)-1]`. Integers in this
+    range *may be* written as JSON integers.
+
+ 1. A big integer of the following form: `{ "#bigint": "[-][0-9]+" }`.
+    We are using this format, as many JSON parsers impose limits on
+    integer values, see [RFC7159][]. Big and small integers *may be*
+    written in this format.
+
+ 1. A list of the form `[ <expr>, ..., <expr> ]`. A list is just a
+    JSON array. TLA+ sequences are written as lists in this format.
+
+ 1. A record of the form `{ "field1": <expr>, ..., "fieldN": <expr> }`. A
+    record is just a JSON object. Field names should not start with `#`
+    and hence should not pose any collision with other constructs.
+    TLA+ records are written as records in this format.
+
+ 1. A tuple of the form `{ "#tup": [ <expr>, ..., <expr> ] }`. There is no
+ strict rule about when to use sequences or tuples. Apalache differentiates
+ between tuples and sequences and it may produce expressions of the both forms.
```

Comment:
```suggestion
 1. A tuple of the form `{ "#tup": [ <expr>, ..., <expr> ] }`. There is no
 strict rule about when to use sequences or tuples. Apalache differentiates
 between tuples and sequences, and it may produce both forms of expressions.
```